frontend/lib/types/TupleType.cpp

```diff
@@ -208,72 +211,77 @@ QualifiedType TupleType::starType() const {
   return subs_.begin()->second;
 }
 
-const TupleType* TupleType::toValueTuple(Context* context) const {
+const TupleType* TupleType::toValueTuple(Context* context, bool makeConst) const {
   // Is it already a value tuple? If so, return that
   bool allValue = true;
+  bool allConst = true;
   int n = numElements();
   for (int i = 0; i < n; i++) {
     const auto& eltType = elementType(i);
     auto kind = eltType.kind();
     if (kind != QualifiedType::VAR)
       allValue = false;
-
+    allConst &= elementType(i).isConst();
     if (eltType.type() && eltType.type()->isTupleType()) {
       // Conservatively throw off 'allValue' because the nested tuple might
       // have a reference inside it.
       allValue = false;
+      allConst = false;
     }
   }
 
-  if (numElements() == 0 || allValue)
+  if (numElements() == 0 || (allValue && allConst >= makeConst))
```

Comment:
Though accurate, I think it's clearer to write this as (!makeConst || allConst) (which is the standard expansion of "make const implies all const", aka "if we need to do X, is X already done?")